### DISCUSSION TITLE
Any plan about supporting mesh shader？

### DISCUSSION THREAD

**Discussion by ludi1991** [o3de/o3de#12646] [Q&A] [answered]
Mesh shader is an important feature for gpu griven pipeline. It seems that O3de does not support mesh shader yet.Any plans?

**Comment by santorac** [comment]
I don't think there are any specific plans for mesh shaders. There is a prototype gem for meshlet rendering (https://github.com/o3de/o3de/tree/development/Gems/Meshlets), but it uses compute shaders, not mesh shaders. I'm not aware of any opposition to having mesh shaders, I think @o3de/sig-graphics-audio is just focused on other things right now.

**Comment by jeremyong-az** [comment, marked as answer]
Thanks for the request/question!

Mesh shaders can be a great accelerant in a GPU-driven pipeline indeed, but we first need to get to a GPU-driven pipeline in the first place :). Even after you have GPU-driven scene rendering going, there are still a few things needed to make mesh shading worthwhile, with some caveats I'll explain later.

For GPU-driven scene rendering, we need the following:

- [x] Persistent scene entity transforms
- [x] Bindless resource support
- [ ] Material resources exposed via bindless heaps on platforms that support it
- [ ] Persistent bound views
- [ ] Compute frustum culling
- [ ] Compute occlusion culling + depth pyramid
- [ ] DrawIndirect pipeline for shadows/primary-camera

As you can see, there are still quite a few items that need to finish first, although progress is being made. Recently, bindless support was added which required changes to both the shader pipeline as well as the RHI management of shader resource groups. Currently, work is also underway to refactor aspects of the material system, with discussion about changing the abstraction by which SRG resources are accessed (indirection may give flexibility in the binding strategy per-platform).

Mesh shading lets you cull at a granularity finer than per-mesh, which I think should come after we handle the coarser-grained per-mesh/object/LOD culling and drawing. The benefit of mesh shading is that the memory forwarded to the top of the rasterization pipeline stays on chip, but note that the culling you do this way can't be forwarded to secondary views, so if the same meshlets are visible in multiple light frustra and/or the main camera, you run that culling code multiple times (or write to main memory for secondary views). Because of this caveat, and the fact that mesh shading isn't ubiquitous, it hasn't been the immediate priority. In principle, we could use it to also decompress mesh data efficiently, but mesh data compressed with some user-codec isn't directly ingestable by the BLAS build pipeline, so its unclear if we'd ever pursue this.

**Reply by ohmaya** [reply]
If we ignore GPU-Driven Pipeline, supporting Mesh & Task shaders themselves should not require so many things, right?

Mesh & Task shaders just replace GPU's traditional geometry pipeline (VS/Tessellation/GS) with rasterization and PS unchanged, so a forward renderer rendering a single mesh can utilize Mesh & Task shaders.

**Reply by jeremyong-az** [reply]
> Mesh & Task shaders just replace GPU's traditional geometry pipeline (VS/Tessellation/GS) with rasterization and PS unchanged, so a forward renderer rendering a single mesh can utilize Mesh & Task shaders.

Without GPU culling/compression of any sort, the experiments I've done indicate that this performs the same or worse (and with less hardware support) :)

**Reply by moudgils** [reply]
Just to re-iterate that mesh shading support will be hard to add to all back - ends (especially if you consider Metal) and ensure it works correctly across all hw/drivers. Hence atleast at this time it doesnt seem like worth the investment. But this may change in future.
